util/src/main/java/tc/oc/pgm/util/menu/InventoryMenuBuilder.java

```diff
@@ -0,0 +1,111 @@
+package tc.oc.pgm.util.menu;
+
+import com.google.common.collect.Lists;
+import java.util.List;
+import net.kyori.text.Component;
+import net.kyori.text.TextComponent;
+import tc.oc.pgm.util.menu.items.InventoryItem;
+import tc.oc.pgm.util.menu.items.InventoryItemBuilder;
+import tc.oc.pgm.util.menu.items.ItemHolder;
+
+public class InventoryMenuBuilder {
+
+  private final InventoryMenuManager manager;
+  private final int rows;
+  private final List<ItemHolder> items;
+  private Component name;
+
+  /**
+   * Creates a new inventory builder
+   *
+   * @param manager
+   * @param rows the number of rows the inventory should have
+   */
+  public InventoryMenuBuilder(InventoryMenuManager manager, int rows) {
+    this.manager = manager;
+    this.rows = rows;
+    this.name = TextComponent.of("");
+    this.items = Lists.newArrayList();
+  }
+
+  /**
+   * Adds an item to the {@link InventoryMenu} being built by this builder at the next free slot.
+   *
+   * @param builder the builder of the item to add
+   * @return this builder
+   */
+  public InventoryMenuBuilder addItem(InventoryItemBuilder builder) {
+    return addItem(builder.build());
+  }
+
+  /**
+   * Adds an item to the {@link InventoryMenu} at the next free slot
+   *
+   * @param item the item to add
+   * @return this builder
+   */
+  public InventoryMenuBuilder addItem(InventoryItem item) {
+    for (int y = 0; y < rows; y++) {
+      for (int x = 0; x < 9; x++) {
+        boolean matched = false;
+        for (ItemHolder itemHolder : items) {
+          if (itemHolder.x == x && itemHolder.y == y) {
+            matched = true;
+            break;
+          }
+        }
```

Comment:
This is too inefficient

I would also skip the `InventoryMenuBuilder` and just have a static `SettingsInventoryMenu` class. By sub-classing instead of tons of builders we remove a ton of boilerplate code, and it makes the menu's more extendable.

Reply:
See my comment above. Not sure what pattern you're suggesting I replace the builders with.